lib/index.js

```diff
@@ -0,0 +1,322 @@
+// FIXME: should joyent copyright be included here `parseRequest` is mostly
+// copy/paste
+
+/*!
+ * Copyright (c) 2018 Digital Bazaar, Inc. All rights reserved.
+ */
+'use strict';
+
+const assert = require('assert-plus');
+const jsprim = require('jsprim');
+const sprintf = require('util').format;
+const url = require('url');
+const HttpSignatureError = require('./HttpSignatureError');
+
+const api = {};
+
+module.exports = api;
+
+api.createAuthzHeader = ({algorithm, includeHeaders, keyId, signature}) => {
+  assert.optionalString(algorithm);
+  assert.string(keyId);
+  assert.arrayOfString(includeHeaders);
+  assert.string(signature);
+
+  let authzFormat;
+  let authz;
+  if(algorithm) {
+    authzFormat =
+      'Signature keyId="%s",algorithm="%s",headers="%s",signature="%s"';
+    authz = sprintf(
+      authzFormat, keyId, algorithm, includeHeaders.join(' '), signature);
+  } else {
+    authzFormat = 'Signature keyId="%s",headers="%s",signature="%s"';
+    authz = sprintf(
+      authzFormat, keyId, includeHeaders.join(' '), signature);
+  }
+  return authz;
+};
+
+api.createStringToSign = ({includeHeaders, requestOptions} = {}) => {
+  assert.object(requestOptions, 'requestOptions');
+  assert.object(requestOptions.headers, 'requestOption.headers');
+  assert.string(requestOptions.method, 'requestOption.method');
+  assert.string(requestOptions.url, 'requestOption.url');
+  assert.optionalArrayOfString(includeHeaders, 'headers');
+
+  const {headers} = requestOptions;
+
+  if(!getHeader(headers, 'date')) {
+    headers.date = jsprim.rfc1123(new Date());
+  }
```

Comment:
I don't think we want to *always* force the date header, rather, if no headers were specified, we add it. Otherwise, there's no way to use this when you can't add the date header ... which may be a thing when using the library client side.

Reply:
Without the date header the clockSkew checking doesn't work which I don't think we want to break.

Reply:
Well, we may *have* to in order to support client side. I expect that an alternative header to "Date" that uses a more granular time may be used instead. If date isn't present ... we should skip checking clockSkew -- if clockSkew is not set. If it is set, we should throw an error and say Date wasn't present. At least that seems reasonable to me.

Reply:
Isn't date our insurance against replay attack, along with providing something in the way of a nonce?  If a party signs a signing string with no headers, that signature could be used over and over right?

Reply:
Yes, if they don't provide their own nonce.

I see that there's some support in here for `x-date` ... maybe we want to check for that too and use either one when doing clockSkew checks.